File: 648-ReplaceWords/648-ReplaceWords.java
/*
648. Replace Words
Medium

In English, we have a concept called root, which can be followed by some other words to form another longer word - let's call this word successor. For example, the root an, followed by other, which can form another word another.

Now, given a dictionary consisting of many roots and a sentence. You need to replace all the successor in the sentence with the root forming it. If a successor has many roots can form it, replace it with the root with the shortest length.

You need to output the sentence after the replacement.

Example 1:

Input: dict = ["cat", "bat", "rat"]
sentence = "the cattle was rattled by the battery"
Output: "the cat was rat by the bat"

 

Note:

    The input will only have lower-case letters.
    1 <= dict words number <= 1000
    1 <= sentence words number <= 1000
    1 <= root length <= 100
    1 <= sentence words length <= 1000
*/

class Solution {
    public String replaceWords(List<String> dict, String sentence) {
        //create a set for dict and interate over all the words in a sentence to replace
        StringBuilder sb = new StringBuilder();
        Set<String> set = new HashSet<>();
        for(String s : dict)
            set.add(s);
        String[] sentences = sentence.split(" ");
        for (String word : sentences){
            String prefix = "";
            for (int i = 1; i <= word.length(); i++) {
                prefix = word.substring(0, i);
                if (set.contains(prefix)) break;
            }
            sb.append(" " + prefix);
        }
        return sb.deleteCharAt(0).toString();
    }
}

class Solution {
    TrieNode root;
    public String replaceWords(List<String> dict, String sentence) {
        //Approach
        // create a trie with all dict values and traverse while iterating through sentence
        StringBuilder sb = new StringBuilder();
        root = new TrieNode();
        for(String s : dict){
            trieInsert(s);
        }
        String[] sentences = sentence.split(" ");
        for(String s : sentences){
            String search = searchTrie(s);
            sb.append(" ").append(search);
        }
        return sb.deleteCharAt(0).toString();
        
    }
    
    private String searchTrie(String word){
        StringBuilder tempsb = new StringBuilder();
        TrieNode curr = root;
         for(int i = 0; i < word.length(); i++){
             char ch = word.charAt(i);
             tempsb.append(ch);
             TrieNode node = curr.children.get(ch);
             if (node == null)
                 return word;
             if(node.isEnd)
                 return tempsb.toString();
             curr = curr.children.get(ch);
         }
        return word;
    }
    
    private void trieInsert(String word){
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
           if (curr.children.containsKey(c)) {
                curr = curr.children.get(c);
                continue;
            }
            curr.children.put(c,new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
    }
}
class TrieNode{
    Map<Character, TrieNode> children;
    boolean isEnd;
    TrieNode(){
        children = new HashMap<>();
        isEnd = false;
    }
}

File: 648-ReplaceWords/648-ReplaceWords.py
"""

648. Replace Words

In English, we have a concept called root, which can be followed by some other word to form another longer word - let's call this word derivative. For example, when the root "help" is followed by the word "ful", we can form a derivative "helpful".

Given a dictionary consisting of many roots and a sentence consisting of words separated by spaces, replace all the derivatives in the sentence with the root forming it. If a derivative can be replaced by more than one root, replace it with the root that has the shortest length.

Return the sentence after the replacement.

Example 1:

Input: dictionary = ["cat","bat","rat"], sentence = "the cattle was rattled by the battery"
Output: "the cat was rat by the bat"

Example 2:

Input: dictionary = ["a","b","c"], sentence = "aadsfasf absbs bbab cadsfafs"
Output: "a a b c"

Constraints:

    1 <= dictionary.length <= 1000
    1 <= dictionary[i].length <= 100
    dictionary[i] consists of only lower-case letters.
    1 <= sentence.length <= 106
    sentence consists of only lower-case letters and spaces.
    The number of words in sentence is in the range [1, 1000]
    The length of each word in sentence is in the range [1, 1000]
    Every two consecutive words in sentence will be separated by exactly one space.
    sentence does not have leading or trailing spaces.

"""

class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:

        def insert_in_trie(word):
            temp = root
            for ch in word:
                if ch not in temp.children:
                    temp.children[ch] = TNode()
                temp = temp.children[ch]
            temp.is_end = True

        def search(word):
            temp = root
            val = ""
            for ch in word:
                if ch not in temp.children: return word
                val += ch
                temp = temp.children[ch]
                if temp.is_end: return val
            return val if temp.is_end else word        
        root = TNode()
        for word in dictionary:
            insert_in_trie(word)
        
        res = []
        for word in sentence.split(" "):
            if word[0] in root.children:
                res.append(search(word))
            else:
                res.append(word)
        return " ".join(res)

class TNode:
    def __init__(self):
        self.children = {}
        self.is_end = False
